### PR TITLE
gettext support

### PR DESCRIPTION
**Pull request by ksqsf**
`gettext` only affects C++ sources. ~~Run Fcitx5 with `LANG=zh_CN.UTF-8` or `LANG=en_US.UTF-8`, and the option names (from domains `fcitx5` and `fcitx5-macos`) should appear in the specified language.~~ The locale is determined by system settings, not from envvars.

image

image

Please also take a look at my translations.

~~(NOTE: text domains from plugins (e.g. `fcitx5-chinese-addons` and `fcitx5-rime`) are not registered successfully due to hardcoded locale dir, so option names from e.g. Pinyin are still English.)~~ (Addon text domains are hardcoded for now. The text domain should be read from metadata in the future.)

### CODE REVIEW

**Comment by wengxt** [comment]
@ksqsf do you want to use fcitx’s transifex for gettext? We have many scripts to pull, merge, and update pot files

File: assets/po/zh_CN.po [hunk 1]
```diff
@@ -0,0 +1,268 @@
...
+"POT-Creation-Date: 2024-04-04 00:25+0800\n"
+"PO-Revision-Date: 2024-04-04 00:50+0800\n"
+"Last-Translator: ksqsf <i@ksqsf.moe>\n"
+"Language-Team: \n"
+"Language: zh_CN\n"
+"MIME-Version: 1.0\n"
+"Content-Type: text/plain; charset=UTF-8\n"
+"Content-Transfer-Encoding: 8bit\n"
+
+#: /Users/kaima/src/fcitx5/macos/macosfrontend/macosfrontend.h:42
```
Comment (eagleoflqj) [inline assets/po/zh_CN.po:19]: There should be no absolute path

File: src/fcitx.cpp [hunk 2]
```diff
@@ -95,10 +96,26 @@ void Fcitx::setupEnv() {
...
+    if (const char *lang = getenv("LANG")) {
+        std::string val = lang;
+        size_t dot_pos = val.find('.');
+        if (dot_pos != std::string::npos) {
+            val = val.substr(0, dot_pos);
+        }
+        val += ":C";
+        setenv("LANGUAGE", val.c_str(), 0);
+    } else {
+        setenv("LANGUAGE", "en_US:C", 1);
```
Comment (eagleoflqj) [inline src/fcitx.cpp:114]: This branch should follow system locale, like Swift part.

**Review by eagleoflqj — CHANGES_REQUESTED** [review]

File: src/fcitx.cpp (hunk 2 above)
Comment (ksqsf) [reply src/fcitx.cpp]: Changed to `setlocale` which respects `LC_*` envvars. Now use `LC_ALL=<locale>` to test.

Comment (ksqsf) [reply src/fcitx.cpp]: Turns out that `setlocale` doesn't always work (in GUI mode). The current locale is now retrieved from Swift, and then use `setlocale` to normalize it.

**Comment by eagleoflqj** [comment]
> do you want to use fcitx’s transifex for gettext

Yes we'd like to use it later when it is ready for initial release. Currently the default implementation doesn't recognize that option value `N_` needs translating.

**Review by eagleoflqj — APPROVED** [review]
This is a huge step. LGTM!

**Pull request merged**